### PR TITLE
WIP Node beta rep

### PR DESCRIPTION
**Pull request by jhagans3**
Start of a reputation system bases on beta probability density function

---
This change is [Reviewable](https://reviewable.io/reviews/storj/storj/116)

### CODE REVIEW

**Comment by navillasa** [comment]
Reviewed 6 of 6 files at r1.
Review status: all files reviewed, 2 unresolved discussions (waiting on @jhagans3)

---

*[protos/nodereputation/nodereputation.proto, line 31 at r1](https://reviewable.io/reviews/storj/storj/116#-LGR14LQ-ZToDDdilvSR:-LGR14LQ-ZToDDdilvSS:bt84pf6) ([raw file](https://github.com/storj/storj/blob/8d0a69864def1e499e61804d6e26aa2dc8ef7468/protos/nodereputation/nodereputation.proto#L31)):*
> ```protobuf
>     int64 limit = 1;
>     repeated Feature featureOrdering = 2;
>     repeated string excuding = 3;
> ```

*excluding (sp)

---

*[protos/nodereputation/nodereputation.proto, line 111 at r1](https://reviewable.io/reviews/storj/storj/116#-LGR1g_Q881oCuUfsy3B:-LGR1g_Q881oCuUfsy3C:brfgtre) ([raw file](https://github.com/storj/storj/blob/8d0a69864def1e499e61804d6e26aa2dc8ef7468/protos/nodereputation/nodereputation.proto#L111)):*
> ```protobuf
>     NEGITIVE_ONE = 0;
>     NEGITIVE_POINT_FIVE = 1;
>     NEGITIVE_POINT_TWENTY_FIVE = 2;
> ```

*NEGATIVE (sp)

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/116)*

**Comment by jhagans3** [comment]
Review status: 4 of 6 files reviewed, 1 unresolved discussion (waiting on @navillasa)

---

*[protos/nodereputation/nodereputation.proto, line 111 at r1](https://reviewable.io/reviews/storj/storj/116#-LGR1g_Q881oCuUfsy3B:-LGR62UoDbCFdR485lYZ:b-896fix) ([raw file](https://github.com/storj/storj/blob/8d0a69864def1e499e61804d6e26aa2dc8ef7468/protos/nodereputation/nodereputation.proto#L111)):*
<details><summary><i>Previously, navillasa (Natalie Villasana) wrote…</i></summary><blockquote>

*NEGATIVE (sp)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/116)*

File: protos/nodereputation/nodereputation.proto [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    NEGATIVE_ONE = 0;
+    NEGATIVE_POINT_FIVE = 1;
+    NEGATIVE_POINT_TWENTY_FIVE = 2;
+    ZERO = 3;
+    POINT_TWENTY_FIVE = 4;
+    POINT_FIVE = 5;
+    ONE = 6;
+}
+
+
```
Comment (nfarah86) [inline protos/nodereputation/nodereputation.proto]: delete extra space

File: protos/nodereputation/nodereputation.proto [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    UpdateRepValue value = 3;
+}
+
+// NodeReputationUpdataStatus tells if a value was set for a node's feature
+message NodeReputationUpdateStatus {
+    string node_id = 1;
+    Feature feature = 2;
+    UpdateStatus status = 3;
+}
+
```
Comment (nfarah86) [inline protos/nodereputation/nodereputation.proto]: delete extra space

File: storage/sqlite3/nodereputation_test.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+func TestSelectReputation(t *testing.T) {
+	testDatabaseName := "./TestSelectReputation.db"
+	db, _ := StartDB(testDatabaseName)
+
+	CreateNewNodeRecord(db, "Alice")
+
+	_, err := GetRep(db, "Alice")
+	if err != nil {
+		panic(err)
```
Comment (nfarah86) [inline storage/sqlite3/nodereputation_test.go]: may be also a helpful log message as well? before you panic, may be put a log message?. I noticed you have a lot of panics. I won't comment on every one. What are your thoughts?

File: storage/sqlite3/nodereputation.go [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+var UpdateError = errs.Class("reputation update error")
+
+// StartDBError is an error class for errors related to the reputation package
+var StartDBError = errs.Class("reputation start sqlite3 error")
+
+// StartDB starts a sqlite3 database from the file path parameter
+func StartDB(filePath string) (*sql.DB, error) {
+	db, err := sql.Open("sqlite3", filePath)
+	if err != nil {
+		return nil, StartDBError.Wrap(err)
```
Comment (nfarah86) [inline storage/sqlite3/nodereputation.go]: if we can't start the db, should we just exit the program? or write a method to try to try to open it again- after failed x times, close the program?

File: storage/sqlite3/nodereputation.go [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+	}
+
+	err = createReputationTable(db)
+	if err != nil {
+		return nil, StartDBError.Wrap(err)
```
Comment (nfarah86) [inline storage/sqlite3/nodereputation.go]: same thing here- if we can't create a table and there is an error, we should probably exit and record it?

**Pull request closed without merging**